### PR TITLE
feat(Geometry): compute ray AABBtree intersections

### PR DESCRIPTION
**Pull request by francoisbonneau**

### CODE REVIEW

File: include/geode/geometry/aabb.h [hunk 1]
```diff
@@ -128,6 +129,27 @@ namespace geode
...
+         * void operator()( index_t cur_element_box ) ;
+         *
+         * @note cur_element_box1 and cur_element_box2 are the element box
+         * indices that intersect.
+         * @note the operator can be defined to test real intersection between
+         * element in boxes and store the result.
+         */
+        template < class EvalIntersection >
+        void compute_ray_trace_element_bbox_intersections(
+            const InfiniteLine< dimension >& ray,
```
Comment (BotellaA) [inline include/geode/geometry/aabb.h:150]: Maybe an alias in basic_objects like `using Ray = InfiniteLine` could be clearer. It is not an `InifiniteLine` even if it is stored the same way.

File: include/geode/geometry/detail/aabb_impl.h [hunk 2]
```diff
@@ -391,4 +409,114 @@ namespace geode
...
+            box_middle, child_left, child_right );
+
+        ray_trace_intersect_recursive< ACTION >(
+            ray, child_left, element_begin, box_middle, action );
+        ray_trace_intersect_recursive< ACTION >(
+            ray, child_right, box_middle, element_end, action );
+    }
+
+    template < index_t dimension >
+    bool ray_box_intersection( const InfiniteLine< dimension >& ray,
```
Comment (BotellaA) [inline include/geode/geometry/detail/aabb_impl.h:451]: This function and the next ones could be private method from AABB::Impl and implemented in aabb.cpp

File: include/geode/geometry/detail/aabb_impl.h [hunk 3]
```diff
@@ -391,4 +409,114 @@ namespace geode
...
+        const BoundingBox< dimension >& box )
+    {
+        const auto box_center = ( box.min() + box.max() ) / 2.;
+        const auto box_extent = ( box.max() - box.min() ) / 2.;
+
+        // Transform the ray to the aligned-box coordinate system.
+        const auto ray_origin = ray.origin() - box_center;
+        const auto rayDirection = ray.direction();
```
Comment (BotellaA) [inline include/geode/geometry/detail/aabb_impl.h:459]:
```suggestion
        const auto& ray_direction = ray.direction();
```
Is this variable necessary?

File: include/geode/geometry/detail/aabb_impl.h [hunk 4]
```diff
@@ -391,4 +409,114 @@ namespace geode
...
+        ray_trace_intersect_recursive< ACTION >(
+            ray, child_right, box_middle, element_end, action );
+    }
+
+    template < index_t dimension >
+    bool ray_box_intersection( const InfiniteLine< dimension >& ray,
+        const BoundingBox< dimension >& box )
+    {
+        const auto box_center = ( box.min() + box.max() ) / 2.;
+        const auto box_extent = ( box.max() - box.min() ) / 2.;
```
Comment (BotellaA) [inline include/geode/geometry/detail/aabb_impl.h:455]: More like half extent... It depends how extent is defined

File: include/geode/geometry/aabb.h [hunk 5]
```diff
@@ -128,6 +129,27 @@ namespace geode
...
+         * defined like this:
+         * void operator()( index_t cur_element_box ) ;
+         *
+         * @note cur_element_box1 and cur_element_box2 are the element box
+         * indices that intersect.
+         * @note the operator can be defined to test real intersection between
+         * element in boxes and store the result.
+         */
+        template < class EvalIntersection >
+        void compute_ray_trace_element_bbox_intersections(
```
Comment (panquez) [inline include/geode/geometry/aabb.h:149]: is ray trace meaningful ? Is not ray enough ?

File: include/geode/geometry/aabb.h [hunk 6]
```diff
@@ -128,6 +129,27 @@ namespace geode
...
+         * element boxes.
+         * @param[in] ray The ray trace to test.
+         * @param[in] action The functor to run when a box is intersected by the
+         * ray trace.
+         *
+         * @tparam EvalIntersection this functor should have an operator()
+         * defined like this:
+         * void operator()( index_t cur_element_box ) ;
+         *
+         * @note cur_element_box1 and cur_element_box2 are the element box
```
Comment (panquez) [inline include/geode/geometry/aabb.h:143]: what/where are cur_element_box1 and cur_element_box2?

File: include/geode/geometry/detail/aabb_impl.h [hunk 7]
```diff
@@ -391,4 +409,114 @@ namespace geode
...
+    bool line_box_intersection( const Point< dimension >& ray_origin,
+        const geode::Vector< dimension >& rayDirection,
+        const Point< dimension >& box_extent );
+
+    template <>
+    bool line_box_intersection( const Point3D& ray_origin,
+        const geode::Vector3D& rayDirection,
+        const Point3D& box_extent )
+    {
+        const auto WxD = geode::Vector3D{ rayDirection }.cross( ray_origin );
```
Comment (BotellaA) [inline include/geode/geometry/detail/aabb_impl.h:482]: What is WxD?

File: include/geode/geometry/detail/aabb_impl.h [hunk 8]
```diff
@@ -391,4 +409,114 @@ namespace geode
...
+        absl::FixedArray< double > absWdU{ std::fabs( rayDirection.value( 0 ) ),
```
Comment (BotellaA) [inline include/geode/geometry/detail/aabb_impl.h:483]: When you know the size at compile time, always use std::array

File: include/geode/geometry/detail/aabb_impl.h [hunk 9]
```diff
@@ -391,4 +409,114 @@ namespace geode
...
+        // Prune sub-tree that does not have intersection
+        if( !ray_box_intersection( ray, node( node_index ) ) )
+        {
+            return;
+        }
+
+        // Leaf case
+        if( is_leaf( element_begin, element_end ) )
+        {
+            // @todo Check if the box is not intersecting itself
```
Comment (panquez) [inline include/geode/geometry/detail/aabb_impl.h:435]: copy/paste artifact? I do not understand this comment

File: include/geode/geometry/detail/aabb_impl.h (hunk 8 above)
Comment (BotellaA) [reply include/geode/geometry/detail/aabb_impl.h:483]: In this case you could also use a Point3D

Comment (BotellaA) [reply include/geode/geometry/detail/aabb_impl.h:483]: And find a better name... if possible

File: include/geode/geometry/detail/aabb_impl.h [hunk 10]
```diff
@@ -391,4 +409,114 @@ namespace geode
...
+    bool line_box_intersection( const Point2D& ray_origin,
+        const geode::Vector2D& rayDirection,
+        const Point2D& box_extent )
+    {
+        const auto lhs =
+            std::fabs( geode::dot_perpendicular( rayDirection, ray_origin ) );
+        const auto rhs =
+            box_extent.value( 0 ) * std::fabs( rayDirection.value( 1 ) )
+            + box_extent.value( 1 ) * std::fabs( rayDirection.value( 0 ) );
+        return ( lhs <= rhs );
```
Comment (BotellaA) [inline include/geode/geometry/detail/aabb_impl.h:520]:
```suggestion
        return lhs <= rhs;
```

File: include/geode/geometry/detail/aabb_impl.h [hunk 11]
```diff
@@ -391,4 +409,114 @@ namespace geode
...
+                  + box_extent.value( 1 ) * absWdU[0] )
+        {
+            return false;
+        }
+
+        return true;
+    }
+    template <>
+    bool line_box_intersection( const Point2D& ray_origin,
+        const geode::Vector2D& rayDirection,
```
Comment (BotellaA) [inline include/geode/geometry/detail/aabb_impl.h:512]:
```suggestion
        const geode::Vector2D& ray_direction,
```
Why not using directly the Ray object? Instead of passing two arguments.

File: tests/geometry/test-aabb.cpp [hunk 12]
```diff
@@ -252,18 +260,115 @@ void test_intersections_with_query_box()
...
+    const index_t nb_boxes{ 10 };
+    const double box_size{ 0.5 };
+    const auto box_vector =
+        create_box_vector< dimension >( nb_boxes, box_size );
+    AABBTree< dimension > aabb( box_vector );
+
+    RayAABBIntersection< dimension > eval_intersection{ box_vector };
+
+    Vector< dimension > ray_direction;
+    ray_direction.set_value( 1, 1.0 );
```
Comment (panquez) [inline tests/geometry/test-aabb.cpp:310]: May you add checks with an oblique ray (e.g. Vector{{0,0,{1,1}} )?

File: include/geode/geometry/aabb.h (hunk 5 above)
Comment (francoisbonneau) [reply include/geode/geometry/aabb.h:149]: any suggestion?

File: include/geode/geometry/aabb.h (hunk 6 above)
Comment (francoisbonneau) [reply include/geode/geometry/aabb.h:143]: i have to update the documentation

File: include/geode/geometry/aabb.h (hunk 5 above)
Comment (panquez) [reply include/geode/geometry/aabb.h:149]: `compute_ray_bbox_intersections`?

File: include/geode/geometry/detail/aabb_impl.h [hunk 13]
```diff
@@ -391,4 +408,115 @@ namespace geode
...
+                return false;
+            }
+        }
+        return line_box_intersection(
+            ray_origin, ray.direction(), box_half_extent );
+    }
+
+    template < index_t dimension >
+    bool line_box_intersection( const Point< dimension >& ray_origin,
+        const geode::Vector< dimension >& ray_direction,
```
Comment (BotellaA) [inline include/geode/geometry/detail/aabb_impl.h:472]: Why did you change the parameters for a Ray?

File: include/geode/geometry/detail/aabb_impl.h [hunk 14]
```diff
@@ -453,70 +453,87 @@ namespace geode
...
+                    > global_epsilon
+                && ray_translated_origin.value( i ) * ray.direction().value( i )
+                       >= global_epsilon )
             {
                 return false;
             }
         }
-        return line_box_intersection(
-            ray_origin, ray.direction(), box_half_extent );
+        return line_box_intersection( ray, box );
```
Comment (BotellaA) [inline include/geode/geometry/detail/aabb_impl.h:469]: in the previous implementation you gave the translated ray as paremeter but not anymore. Why not giving the translated line?

File: include/geode/geometry/aabb.h [hunk 15]
```diff
@@ -101,11 +102,8 @@ namespace geode
          * defined like this:
          * void operator()( index_t cur_element_box ) ;
          *
-         * @note where cur_element_box is the element box index
-         * that is intersected by \p box.
-         * @note A posibility is to define this operator to add \p
-         * cur_element_box to a set and build a set of tree element that
-         * intersect the \p box.
+         * @note the operator define what to do with the box \p cur_element_box
```
Comment (panquez) [inline include/geode/geometry/aabb.h:105]:
```suggestion
         * @note the operator defines what to do with the box \p cur_element_box
```

File: include/geode/geometry/aabb.h [hunk 16]
```diff
@@ -121,13 +119,31 @@ namespace geode
          * ;
          * @note cur_element_box1 and cur_element_box2 are the element box
          * indices that intersect.
-         * @note the operator can be defined to test real intersection between
-         * element in boxes and store the result.
+         * @note the operator defines what to do when two boxes of the
+         * tree ( \p cur_element_box1 and \p cur_element_box2 ) intesect each
```
Comment (panquez) [inline include/geode/geometry/aabb.h:123]:
```suggestion
         * tree ( \p cur_element_box1 and \p cur_element_box2 ) intersect each
```

File: include/geode/geometry/aabb.h [hunk 17]
```diff
@@ -121,13 +119,31 @@ namespace geode
...
+        /*!
+         * @brief Computes the intersections between a given ray and all
+         * element boxes.
+         * @param[in] ray The ray to test.
+         * @param[in] action The functor to run when a box is intersected by the
+         * ray.
+         * @tparam EvalIntersection this functor should have an operator()
+         * defined like this:
+         * void operator()( index_t cur_element_box ) ;
+         * @note the operator define what to do with the box \p cur_element_box
```
Comment (panquez) [inline include/geode/geometry/aabb.h:140]:
```suggestion
         * @note the operator defines what to do with the box \p cur_element_box
```

File: include/geode/geometry/detail/aabb_impl.h [hunk 18]
```diff
@@ -391,4 +408,132 @@ namespace geode
...
+    template < index_t dimension >
+    template < typename ACTION >
+    void AABBTree< dimension >::Impl::ray_intersect_recursive(
+        const Ray< dimension >& ray,
+        index_t node_index,
+        index_t element_begin,
+        index_t element_end,
+        ACTION& action ) const
+    {
+        OPENGEODE_ASSERT( node_index < tree_.size(), "Node out of tree range" );
```
Comment (panquez) [inline include/geode/geometry/detail/aabb_impl.h:421]:
```suggestion
        OPENGEODE_ASSERT( node_index < tree_.size(), "[AABBTree::ray_intersect_recursive] Node out of tree range" );
```

File: include/geode/geometry/detail/aabb_impl.h [hunk 19]
```diff
@@ -391,4 +408,132 @@ namespace geode
...
+        OPENGEODE_ASSERT(
+            element_begin != element_end, "No iteration allowed start == end" );
```
Comment (panquez) [inline include/geode/geometry/detail/aabb_impl.h:423]:
```suggestion
            element_begin != element_end, "[AABBTree::ray_intersect_recursive] No iteration allowed start == end" );
```

File: include/geode/geometry/detail/aabb_impl.h [hunk 20]
```diff
@@ -391,4 +408,132 @@ namespace geode
...
+    {
+        const auto box_center = ( box.min() + box.max() ) / 2.;
+        const auto box_half_extent = ( box.max() - box.min() ) / 2.;
+
+        // Transform the ray to the aligned-box coordinate system.
+        const auto line_translated_origin = line.origin() - box_center;
+
+        const auto orign_cross_direction =
+            line.direction().cross( line_translated_origin );
+        geode::Point3D abs_line_direction = {
```
Comment (panquez) [inline include/geode/geometry/detail/aabb_impl.h:487]:
```suggestion
        const geode::Point3D abs_line_direction = {
```

File: tests/geometry/test-aabb.cpp [hunk 21]
```diff
@@ -148,12 +151,14 @@ void test_nearest_neighbor_search()
                 aabb.closest_element_box( query, disteval );
 
             OPENGEODE_EXCEPTION( box_id == global_box_index( i, j, nb_boxes ),
-                "[Test] Wrong nearest box index" );
+                "[Test]  Nearest box to point AABB - Wrong nearest box index" );
```
Comment (panquez) [inline tests/geometry/test-aabb.cpp:154]:
```suggestion
                "[Test] Nearest box to point AABB - Wrong nearest box index" );
```

File: tests/geometry/test-aabb.cpp [hunk 22]
```diff
@@ -148,12 +151,14 @@ void test_nearest_neighbor_search()
...
             OPENGEODE_EXCEPTION( nearest_point == box_center,
-                "[Test] Wrong nearest box center " );
+                "[Test]  Nearest box to point AABB - Wrong nearest box "
```
Comment (panquez) [inline tests/geometry/test-aabb.cpp:156]:
```suggestion
                "[Test] Nearest box to point AABB - Wrong nearest box "
```

File: tests/geometry/test-aabb.cpp [hunk 23]
```diff
@@ -148,12 +151,14 @@ void test_nearest_neighbor_search()
...
+                "center " );
             const Vector< dimension > vec{ box_center, query };
             OPENGEODE_EXCEPTION( distance == vec.length(),
-                "[Test] Wrong distance to nearest box center" );
+                "[Test]  Nearest box to point AABB - Wrong distance to nearest "
```
Comment (panquez) [inline tests/geometry/test-aabb.cpp:160]:
```suggestion
                "[Test] Nearest box to point AABB - Wrong distance to nearest "
```

File: tests/geometry/test-aabb.cpp [hunk 24]
```diff
@@ -229,7 +236,8 @@ void test_intersections_with_query_box()
 
             OPENGEODE_EXCEPTION(
                 eval_intersection.box_intersections_.size() == 4,
-                "[Test] Wrong number of intersected boxes" );
+                "[Test]  Box-Box intersection - Wrong number of intersected "
```
Comment (panquez) [inline tests/geometry/test-aabb.cpp:239]:
```suggestion
                "[Test] Box-Box intersection - Wrong number of intersected "
```

File: tests/geometry/test-aabb.cpp [hunk 25]
```diff
@@ -252,18 +260,144 @@ void test_intersections_with_query_box()
...
+    Vector< dimension > ray_direction;
+    ray_direction.set_value( 1, 1.0 );
+
+    for( const index_t i : Range( nb_boxes ) )
+    {
+        Point< dimension > ray_origin;
+        ray_origin.set_value( 0, i );
+        ray_origin.set_value( 1, i );
+
+        Ray< dimension > query{ ray_direction, ray_origin };
```
Comment (panquez) [inline tests/geometry/test-aabb.cpp:318]:
```suggestion
        const Ray< dimension > query{ ray_direction, ray_origin };
```

File: tests/geometry/test-aabb.cpp [hunk 26]
```diff
@@ -252,18 +260,144 @@ void test_intersections_with_query_box()
...
+        eval_intersection.box_intersections_.clear();
+        aabb.compute_ray_element_bbox_intersections( query, eval_intersection );
+
+        OPENGEODE_EXCEPTION(
+            eval_intersection.box_intersections_.size() == nb_boxes - i,
+            "[Test] Box-Ray intersection - Wrong number of boxes intersected "
+            "by the ray " );
+
+        absl::flat_hash_set< index_t > expected_set;
+        for( const index_t c : Range( nb_boxes - i ) )
```
Comment (panquez) [inline tests/geometry/test-aabb.cpp:329]:
```suggestion
        for( const auto c : Range{ nb_boxes - i } )
```

File: tests/geometry/test-aabb.cpp [hunk 27]
```diff
@@ -252,18 +260,144 @@ void test_intersections_with_query_box()
...
+        OPENGEODE_EXCEPTION(
+            eval_intersection.box_intersections_ == expected_set,
+            "[Test] Box-Ray intersection - Wrong set of boxes" );
+    }
+    // ray between two boxes
+    Point< dimension > ray_origin;
+    ray_origin.set_value( 0, box_size );
+    ray_origin.set_value( 1, box_size );
+
+    Ray< dimension > query{ ray_direction, ray_origin };
```
Comment (panquez) [inline tests/geometry/test-aabb.cpp:343]:
```suggestion
    const Ray< dimension > query{ ray_direction, ray_origin };
```

File: tests/geometry/test-aabb.cpp [hunk 28]
```diff
@@ -252,18 +260,144 @@ void test_intersections_with_query_box()
...
+
+    eval_intersection.box_intersections_.clear();
+    aabb.compute_ray_element_bbox_intersections( query, eval_intersection );
+
+    OPENGEODE_EXCEPTION(
+        eval_intersection.box_intersections_.size() == 2 * nb_boxes,
+        "[Test] Box-Ray intersection - Wrong number of boxes intersected by "
+        "the ray " );
+    absl::flat_hash_set< index_t > expected_set;
+    for( const index_t c : Range( nb_boxes ) )
```
Comment (panquez) [inline tests/geometry/test-aabb.cpp:353]:
```suggestion
    for( const auto c : Range{ nb_boxes } )
```

File: tests/geometry/test-aabb.cpp [hunk 29]
```diff
@@ -252,18 +260,144 @@ void test_intersections_with_query_box()
...
+        "[Test] Box-Ray intersection - Wrong set of boxes" );
+
+    // oblique ray
+    ray_direction.set_value( 0, 1.0 );
+    ray_direction.set_value( 1, 1.0 );
+
+    ray_origin.set_value( 0, box_size );
+    ray_origin.set_value( 1, box_size );
+
+    Ray< dimension > query2{ ray_direction, ray_origin };
```
Comment (panquez) [inline tests/geometry/test-aabb.cpp:368]:
```suggestion
    const Ray< dimension > query2{ ray_direction, ray_origin };
```

File: tests/geometry/test-aabb.cpp [hunk 30]
```diff
@@ -252,18 +260,144 @@ void test_intersections_with_query_box()
...
+
+    OPENGEODE_EXCEPTION(
+        eval_intersection.box_intersections_.size() == 3 * ( nb_boxes - 1 ) + 1,
+        "[Test] Box-Ray intersection - Wrong number of boxes intersected by "
+        "the ray " );
+
+    expected_set.clear();
+    expected_set.emplace( global_box_index( 0, 0, nb_boxes ) );
+    expected_set.emplace( global_box_index( 1, 0, nb_boxes ) );
+    for( const index_t c : Range( nb_boxes - 2 ) )
```
Comment (panquez) [inline tests/geometry/test-aabb.cpp:381]:
```suggestion
    for( const auto c : Range{ nb_boxes - 2 } )
```
